Implemented Accounts and Account Messages

acedward opened the pull request:
# Account Commands
<img width="1086" height="708" alt="image" src="https://github.com/user-attachments/assets/439ecb27-6475-4018-ae12-f06a9e41debc" />

## Commands
```
"&createAccount", [
],

"&linkAddress", [
    ["account_id", Type.Number()],
    ["signature_from_primary", Type.String()],
    ["new_address", Type.String()],
    ["signature_from_new_address", Type.String()],
    ["is_new_primary", Type.Boolean()],
  ],

"&unlinkAddress", [
    ["account_id", Type.Number()],
    // if empty, unlink self.
    ["signature_from_primary", Type.String()],
    ["account_address", Type.String()],
    ["new_primary", Type.String()],
  ],
```

## Signed Messages

**Link**
`link:${String(account_id)}:${other_address}:${
      is_new_primary ? "true" : "false"
    }`

**Unlink**
`unlink:${String(account_id)}:${other_address}:${new_primary ?? ""}`
```

acedward committed:
Implemented Accounts and Account Messages

SebastienGllmt commented on packages/paima-sdk/concise/src/delegate.ts:
```diff
@@ -0,0 +1,124 @@
+  const account = privateKeyToAccount(privateKey);
+  const walletClient = createWalletClient({
+    account,
+    transport: http("http://0.0.0.0"),
```
TODO

acedward replied:
viem requires with this method a http url or gives a warning in the terminal, this is also not required if chain is defined as it has a defined url

SebastienGllmt commented on packages/node-sdk/sm/primitives/evm/rpc/paima-l2.ts:
```diff
@@ -37,75 +77,73 @@ function* executePaimaL2Input(input: {
-  // const primitiveName = response.output.syncProtocol.payload.primitiveName;
+}): StateUpdateStream<void> {
+  // For EVM we use the lowercase address.
+  const normalizedSignerAddress = input.signerAddress
```
you can manually call lowercase and then cast, but fyi you can also do this through Typebox encode/decode via `TypeboxHelpers.Evm.Address`

SebastienGllmt commented on packages/node-sdk/sm/primitives/evm/rpc/paima-l2.ts:
```diff
@@ -11,19 +11,59 @@ import type { StateUpdateStream } from "@paima/coroutine";
+  nonce: string | undefined,
+  block_height: BlockNumber,
+): StateUpdateStream<boolean> {
+  // TODO This is only for batched messages?
```
TODO?

acedward replied:
PAI-56 (I only moved this todo in this PR)

<img width="1185" height="794" alt="image" src="https://github.com/user-attachments/assets/3567ba8b-8814-488d-9a98-426825774297" />

SebastienGllmt commented on packages/paima-sdk/crypto/src/chains/polkadot.ts:
```diff
@@ -0,0 +1,28 @@
+//       return false;
+//     }
+//   };
+// }
```
```suggestion
import { doLog, Signature, SubstrateAddress, TypeboxHelpers } from '@paima/utils';
import type { IVerify } from './IVerify.js';
import { Value } from '@sinclair/typebox/value';

export class PolkadotCrypto implements IVerify {
  verifyAddress = (address: string): address is SubstrateAddress => {
    return Value.Check(TypeboxHelpers.Substrate.Address, address);
  };
  verifySignature = async (
    userAddress: SubstrateAddress,
    message: string,
    signature: Signature
  ): Promise<boolean> => {
    try {
      const { cryptoWaitReady, decodeAddress, signatureVerify } = await import(
        '@polkadot/util-crypto'
      );
      const { u8aToHex } = await import('@polkadot/util');
      await cryptoWaitReady();
      const publicKey = decodeAddress(userAddress);
      const hexPublicKey = u8aToHex(publicKey);
      return signatureVerify(message, signature, hexPublicKey).isValid;
    } catch (err) {
      doLog('[address-validator] error verifying polkadot signature:', err);
      return false;
    }
  };
}
```

SebastienGllmt commented on packages/paima-sdk/crypto/src/chains/mina.ts:
```diff
@@ -0,0 +1,45 @@
+//       return false;
+//     }
+//   };
+// }
```
```suggestion
import { doLog, MinaAddress, Signature, TypeboxHelpers } from '@paima/utils';
import type { IVerify } from './IVerify.js';
import type { NetworkId } from 'mina-signer';
import { Value } from '@sinclair/typebox/value';

export class MinaCrypto implements IVerify {
  verifyAddress = (address: string): address is MinaAddress => {
    return Value.Check(TypeboxHelpers.Mina.Address, address);
  };
  verifySignature = async (
    userAddress: MinaAddress,
    message: string,
    sigStruct: Signature
  ): Promise<boolean> => {
    try {
      const [field, scalar, network, ...remainder] = sigStruct.split(';');
      if (!field || !scalar || !network || remainder.length > 0) {
        return false;
      }

      const Client = (await import('mina-signer')).default;

      const signerClient = new Client({ network: network as NetworkId });

      const verifyBody = {
        data: message,
        publicKey: userAddress,
        signature: { field, scalar },
      };

      const verifyResult = signerClient.verifyMessage(verifyBody);

      return verifyResult;
    } catch (err) {
      doLog('[address-validator] error verifying mina signature:', err);
      return false;
    }
  };
}
```

SebastienGllmt commented on packages/paima-sdk/crypto/src/chains/evm.ts:
```diff
@@ -0,0 +1,26 @@
+export class EvmCrypto implements IVerify {
+  verifyAddress = (address: string): Promise<boolean> => {
+    return new Promise((resolve) => resolve(isAddress(address)));
+  };
```
```suggestion
  verifyAddress = (address: string): address is EvmAddress => {
    return Value.Check(TypeboxHelpers.Evm.Address, address);
  };
```

SebastienGllmt commented on packages/paima-sdk/crypto/src/chains/evm.ts:
```diff
@@ -0,0 +1,26 @@
+    return new Promise((resolve) => resolve(isAddress(address)));
+  };
+  verifySignature = async (
+    signerAddress: `0x${string}`,
+    message: string,
+    signature: `0x${string}`,
```
```suggestion
    signerAddress: EvmAddress,
    message: string,
    signature: Signature
```

SebastienGllmt commented on packages/paima-sdk/crypto/src/chains/cardano.ts:
```diff
@@ -0,0 +1,28 @@
+// import { doLog } from '@paima/utils';
+// import type { IVerify } from './IVerify.js';
+
+// export class CardanoCrypto implements IVerify {
+//   verifyAddress = async (address: string): Promise<boolean> => {
+//     // TODO: improve
+//     return await Promise.resolve(/^addr1[a-zA-Z0-9]+$/.test(address));
+//   };
+//   verifySignature = async (
+//     userAddress: string,
+//     message: string,
+//     sigStruct: string
+//   ): Promise<boolean> => {
+//     try {
+//       const [signature, key, ...remainder] = sigStruct.split('+');
+//       if (!signature || !key || remainder.length > 0) {
+//         return false;
+//       }
+//       const { default: verifyCardanoDataSignature } = await import(
+//         '@cardano-foundation/cardano-verify-datasignature'
+//       );
+//       return verifyCardanoDataSignature(signature, key, message, userAddress);
+//     } catch (err) {
+//       doLog('[address-validator] error verifying cardano signature:', err);
+//       return false;
+//     }
+//   };
+// }
```
```suggestion
import { CardanoAddress, doLog, Signature, TypeboxHelpers } from '@paima/utils';
import type { IVerify } from './IVerify.js';
import { Value } from '@sinclair/typebox/value';

export class CardanoCrypto implements IVerify {
  verifyAddress = (address: string): address is CardanoAddress => {
    return Value.Check(TypeboxHelpers.Cardano.Address, address);
  };
  verifySignature = async (
    userAddress: CardanoAddress,
    message: string,
    sigStruct: Signature
  ): Promise<boolean> => {
    try {
      const [signature, key, ...remainder] = sigStruct.split('+');
      if (!signature || !key || remainder.length > 0) {
        return false;
      }
      const { default: verifyCardanoDataSignature } = await import(
        '@cardano-foundation/cardano-verify-datasignature'
      );
      return verifyCardanoDataSignature.default(signature, key, message, userAddress);
    } catch (err) {
      doLog('[address-validator] error verifying cardano signature:', err);
      return false;
    }
  };
}
```

SebastienGllmt commented on packages/paima-sdk/crypto/src/chains/algorand.ts:
```diff
@@ -0,0 +1,72 @@
+//     const { decodeSignedTransaction } = await import('algosdk');
+//     return decodeSignedTransaction(signedTx);
+//   };
+// }
```
```suggestion
import { CardanoAddress, doLog, Signature, TypeboxHelpers } from '@paima/utils';
import type { IVerify } from './IVerify.js';
import { Value } from '@sinclair/typebox/value';

export class CardanoCrypto implements IVerify {
  verifyAddress = (address: string): address is CardanoAddress => {
    return Value.Check(TypeboxHelpers.Cardano.Address, address);
  };
  verifySignature = async (
    userAddress: CardanoAddress,
    message: string,
    sigStruct: Signature
  ): Promise<boolean> => {
    try {
      const [signature, key, ...remainder] = sigStruct.split('+');
      if (!signature || !key || remainder.length > 0) {
        return false;
      }
      const { default: verifyCardanoDataSignature } = await import(
        '@cardano-foundation/cardano-verify-datasignature'
      );
      return verifyCardanoDataSignature.default(signature, key, message, userAddress);
    } catch (err) {
      doLog('[address-validator] error verifying cardano signature:', err);
      return false;
    }
  };
}
```

SebastienGllmt commented on packages/paima-sdk/concise/src/delegate.ts:
```diff
@@ -0,0 +1,124 @@
+    return [BuiltinGrammarPrefix.createAccount];
+  },
+  linkAddress: async (
+    primaryAccountPrivateKey: `0x${string}`,
```
we should try and use nominal types (`FastFlavor`) for these wherever possible

acedward committed:
Refactor to e2e test to keep shared state

acedward committed:
Using Crypto Validation for Batcher Messages

acedward committed:
Update packages/paima-sdk/crypto/src/chains/polkadot.ts

Co-authored-by: Sebastien Guillemot <SebastienGllmt@gmail.com>

acedward committed:
Update packages/paima-sdk/crypto/src/chains/mina.ts

Co-authored-by: Sebastien Guillemot <SebastienGllmt@gmail.com>

acedward committed:
Update packages/paima-sdk/crypto/src/chains/cardano.ts

Co-authored-by: Sebastien Guillemot <SebastienGllmt@gmail.com>

acedward committed:
Update packages/paima-sdk/crypto/src/chains/algorand.ts

Co-authored-by: Sebastien Guillemot <SebastienGllmt@gmail.com>

acedward committed:
Update packages/paima-sdk/crypto/src/chains/evm.ts

Co-authored-by: Sebastien Guillemot <SebastienGllmt@gmail.com>

acedward committed:
Update packages/paima-sdk/crypto/src/chains/evm.ts

Co-authored-by: Sebastien Guillemot <SebastienGllmt@gmail.com>

acedward committed:
Added Crypto Managers

acedward committed:
Update explorer for accounts.
Added rule that main cannot unlink without selecting a new main.
Updated file wallet-delegate -> accounts
Better Types

acedward committed:
Clean up more types

acedward committed:
Cleaned up types

acedward committed:
Clean up types

acedward commented on packages/paima-sdk/crypto/src/chains/algorand.ts:
```diff
@@ -0,0 +1,44 @@
+    userAddress: WalletAddress,
+    message: string,
+    sigStruct: Signature,
+  ): Promise<boolean> => {
```
Is this OK? Same validation as Cardano?

acedward committed:
Clean up types

acedward committed:
Basic Wallet Management in Explorer

acedward committed:
Merge remote-tracking branch 'origin/v-next' into feature/v-next-accounts

SebastienGllmt commented on example/client/node/scripts/e2e-contracts.ts:
```diff
@@ -18,17 +18,65 @@ import {
+  address: `0x${string}`;
+  privateKey: `0x${string}`;
+}[] = [{
+  address: "0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266",
```
if these come from hardhat, we should mention this

SebastienGllmt commented on example/client/node/scripts/e2e-shared-state.ts:
```diff
@@ -0,0 +1,117 @@
+
+export function addLinkedAddress(
+  sharedState: SharedState,
+  _address: string,
```
prefer to use nominal types when possible (see `nominal.ts`)

I recommend this for everything in this PR, otherwise it will get impossible to track what type things actually are

SebastienGllmt commented on packages/paima-sdk/crypto/src/chains/polkadot.ts:
```diff
@@ -0,0 +1,35 @@
+    signature: Signature,
+  ): Promise<boolean> => {
+    try {
+      const { cryptoWaitReady, decodeAddress, signatureVerify } = await import(
```
FYI I think for this stuff we'll be able to use `import defer` (which isn't in Deno yet, but will probably come sooner than later)
